### PR TITLE
Add Strict Metaschema Validation

### PR DESCRIPTION
**Pull request by holmanb**
## Proposed Commit Message

```
Improve schema validation.

This adds strict validation of config module definitions at testing
time, with plumbing included for future runtime validation. This
eliminates a class of bugs resulting from schemas that have definitions
that are incorrect, but get interpreted by jsonschema as
"additionalProperties" that are therefore ignored.


- Add strict meta-schema for jsonschema unit test validation
- Separate schema from module metadata structure
- Improve type annotations for various functions and data types

Cleanup:
- Remove unused jsonschema "required" elements 
- Eliminate manual memoization in schema.py:get_schema(),
        reference module.__doc__ directly
```

## Test Steps
Many of the files were only touched for function name/arg/schema/meta updates, however schema.py, test_schema.py, test_cli.py had the most substantial changes.

## Checklist:

 - [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/topics/hacking.html)
 - [x] I have updated or added any unit tests accordingly
 - [x] I have updated or added any documentation accordingly

### CODE REVIEW

File: tests/unittests/test_handler/test_schema.py
Comment (holmanb) [inline tests/unittests/test_handler/test_schema.py:613]: This line is basically the only difference between an otherwise identical function in `schema.py`. I'm leaning towards adding a flag to the function signature in schema.py to set additionalProperties=False, which would reduce this diff by about 70 LOC.

File: doc/rtd/conf.py
Comment (blackboxsw) [inline doc/rtd/conf.py:75]: Minimally I think we need a split("\n") here because you are appending a multi-line string to lines (list). 

The current branch doens't show any module names in the bulleted list: `tox -e doc; xdg-open doc/rtd_html/topics/modules.html`

**Comment by blackboxsw** [comment]
Thanks for this @holmanb. I'm just starting the review amd will have some significant feedback tomorrow.  In your suggested commit message could you capture the intent behind the strict_metaschema being a testing time validation of config module definitions rather than runtime validation.

File: cloudinit/config/schema.py
Comment (holmanb) [inline cloudinit/config/schema.py:357]: avoid builtin function name, replace with something more specific

File: cloudinit/config/schema.py
Comment (holmanb) [inline cloudinit/config/schema.py:323]: If SCHEMA_UNDEFINED can be of type `str` rather than type `bytes`, this annotation would be simpler and the changes below would not be necessary. On line 307 if `property_dict.get('enum')` returns a falsy value, then SCHEMA_UNDEFINED is returned from this function unmodified as type `bytes`. Alternatively casting `property_type` to `str` would also accomplish basically the same thing.

File: tests/unittests/test_cli.py
Comment (blackboxsw) [inline tests/unittests/test_cli.py:237]: This line may to cause problems openEuler vs. openeuler with #1117 fix I think.

File: tests/unittests/test_cli.py
Comment (blackboxsw) [inline tests/unittests/test_cli.py:222]: Since we are touching this line, let's update inclusive wording where we can.
```suggestion
        """Validate that doc content from known examples is well-structured.
```

or coherent or correct etc.

File: cloudinit/config/schema.py
Comment (holmanb) [inline cloudinit/config/schema.py:413]: This global was previously a reimplementation of features provided by standard library functionality (`lru_cache`). The new implementation uses `load_doc()`, which references the docstrings that are already computed on module load. This means that, like the previous implementation, docstrings are only generated once.

File: tests/unittests/test_handler/test_schema.py
Comment (holmanb) [inline tests/unittests/test_handler/test_schema.py:542]: This test provides test-time validation of schemas.

File: tests/unittests/test_cli.py
Comment (blackboxsw) [inline tests/unittests/test_cli.py:227]: Ultimately, we'd like to move away from subclassing test_helpers.FilesystemMockingTestCase which is a subclass of unittest.TestCase derivatives where possible and prefer to use pytest framework for discovering any tests prefaced with "Test". This gives the ability to use capsys fixture. I'm not certain this particular case is a candidate for shifting aware from unittest.TestCase for the other unit tests in this test_cli module. But I can see that the particular test case is an easy shift away from the FileSystemMocking test case. Your call on what you'd prefer to do on this PR. It might be more work than it's worth for this iteration to restructure the tests in this module.

```diff
diff --git a/tests/unittests/test_cli.py b/tests/unittests/test_cli.py
index d6909e7a..64e7b8cb 100644
--- a/tests/unittests/test_cli.py
+++ b/tests/unittests/test_cli.py
@@ -282,4 +282,31 @@ class TestCLI(test_helpers.FilesystemMockingTestCase):
         self.assertFalse(parseargs.debug)
         self.assertFalse(parseargs.force)
 
+
+class TestUsePyTest:
+    def _call_main(self, sysv_args=None):
+        if not sysv_args:
+            sysv_args = ['cloud-init']
+        try:
+            return cli.main(sysv_args=sysv_args)
+        except SystemExit as e:
+            return e.code
+
+    def test_wb_devel_schema_subcommand_doc_content(self, capsys):
+        """Validate that doc content is sane from known examples.
+        Ensure that schema doc is returned
+        """
+        self._call_main(['cloud-init', 'devel', 'schema', '--docs', 'all'])
+        expected_doc_sections = [
+                '**Supported distros:** all',
+                ('**Supported distros:** almalinux, alpine, centos, '
+                    'cloudlinux, debian, eurolinux, fedora, openEuler, '
+                    'opensuse, photon, rhel, rocky, sles, ubuntu, virtuozzo'),
+                '**Config schema**:\n    **resize_rootfs:** '
+                '(true/false/noblock)',
+                '**Examples**::\n\n    runcmd:\n        - [ ls, -l, / ]\n'
+        ]
+        stdout, stderr = capsys.readouterr()
+        for expected in expected_doc_sections:
+            assert expected in stdout
 # : ts=4 expandtab
```

File: tests/unittests/test_handler/test_schema.py
Comment (blackboxsw) [inline tests/unittests/test_handler/test_schema.py:546]: If we don't have one, let's also add a unit test which asserts no invalid module-level schema raises a schema validation error when strict_metaschema=True. maybe with a pytest.mark.parametrize True,False on strict_metaschema to see validation error vs hiding it.

**Review by blackboxsw — COMMENTED** [review]
@holmanb thanks for putting this together. A couple of minor requests inline and a general question for you about moving toward pytest to avoid all the stdout/error context manager work.

File: cloudinit/config/schema.py
Comment (holmanb) [reply cloudinit/config/schema.py:132]: > @holmanb What do you feel is the use-case we are solving where we want the early exit and ignore of the schema validation below if strict_metaschema=True?

My bad, the intention is not clear here.

Originally this was a separate function in the unittest, but I shaved about 70LOC from the diff by making it a single `if` statement in `schema.py`. I'm realizing that perhaps refactoring out a separate function `get_schema_validator()` for use in `validate_cloudconfig_schema()` and the unittest would probably have been better than muddying the purpose of `validate_cloudconfig_schema()` by making it dual-purpose.

Basically the logic is currently:
```
if strict_metaschema:
    <validate the schema using the metaschema>
    return
<validate the config using the schema>
```

maybe clearer would have been:
```
if strict_metaschema:
    <validate schema using metaschema>
else:
    <validate config using schema>
```
The unittest passes an empty config when calling `validate_cloudconfig_schema()`, since the purpose of that tests is to validate the schemas, not to validate a config with the schema.

> And is it worth a unittest to show that error conditions in schema don't get raised with strict_metaschema=True. (such a unit test docstring would also be a place to document why we don't want to raise errors at runtime).

This sounds like a valuable addition. I will add that.

**Review by TheRealFalcon — COMMENTED** [review]
A few inline questions/comments, but overall looking good.

File: tests/unittests/test_cli.py
Comment (holmanb) [reply tests/unittests/test_cli.py:227]: I haven't used capsys directly before, but it looks like `pytest --capture=tee-sys` with capsys would accomplish most of what I was trying to do. This test in particular produces a large amount of output, so the contextmanager allowed me to programatically filter values before printing captured output for debugging, which I don't see as an option with capsys (however maybe I just missed that in the docs). For most use cases duplicating output to stdout/stderr would be more than enough.

I am happy to separate this case out using the suggestion above, but given the size of this PR already, I think I'd prefer to make a `s/FileSystemMocking/capsys/g` conversion of this file into a separate PR.

File: doc/rtd/conf.py
Comment (holmanb) [reply doc/rtd/conf.py:72]: My lsp server (pyright I think) didn't recognize the `def generate_docstring_from_schema(_app, what, _name, obj, _options, lines):` syntax, but it didn't mind what I did above. A PR/bug with my lsp server would be more appropriate than working around it in our code, sorry about that.

File: cloudinit/config/schema.py
Comment (blackboxsw) [inline cloudinit/config/schema.py:126]: Is this code necessary or referenced anywhere anymore?

File: cloudinit/config/schema.py
Comment (blackboxsw) [reply cloudinit/config/schema.py:132]: @holmanb I'm defintiely all for early exits specifically for avoiding nested scopes too.  It feel clearer in my opinion as we can then avoid additional line-wraps due to overflow due to all the extra indents later in functions/methods. All good there.  I don't really mind the dual-purpose of the validate_cloudconfig_schema(), that seems to make sense to me as we are generally validating different scopes of the "schema".

My comment really was that don't see why the two options need to be mutually-exclusive. If we add a runtime call with `strict_metaschema=True`, I see a use-case where cloud-init may apply strict schema verification during runtime for both meta schema and module config schema. Maybe cloud-init would warn on invalid meta schema defintions (because that only affects docs) and error on invalid "config" schema.  Also our existing unit tests in this branch pass either way with the empty config.

Given that is is a "future" discussion, and we don't have any runtime call-sites passing `strict_metadata=True`, it's definitely not something that needs resolution in this PR.

File: doc/rtd/conf.py
Comment (holmanb) [reply doc/rtd/conf.py:70]: This currently implements similar logic as before but without the memoized function call (since the __doc__ is given the output of `get_meta_doc()` already). Since the docstring already has what it needs I don't think this preprocessing is required at all.

File: tests/unittests/test_cli.py
Comment (blackboxsw) [reply tests/unittests/test_cli.py:227]: +1 this can wait. not needed for this branch.

File: cloudinit/config/schema.py
Comment (blackboxsw) [reply cloudinit/config/schema.py:413]: Docstrings are only generated once, but get_schema is still recalculating what full_schema is each time it is called. If this function was called multiple times throughout cloud-init bringup, we'd probably want to put an `@lru_cache` on  this I think, but since we only have one call site for get_schema() `validate_cloudconfig_file` and it's a manual command on the CLI, I suggest we don't worry about that pre-mature caching that was done.

File: cloudinit/config/schema.py
Comment (blackboxsw) [inline cloudinit/config/schema.py:37]: We have dropped the description of each property here in your commit 859bcc2f8ab371f2bfe19bacc4cd31430e50ece7. This makes both our online docs and the command-line docs a lot less helpful when people try to figure out how to use a given key.

This results in more sparse/terse documentation that no longer explains the intent and expectations of each property value for both commandline docs and online docs as we can see in the [write_files docs](https://cloudinit.readthedocs.io/en/latest/topics/modules.html#write-files)

```
content: (string) Optional content to write to the provided path. When content is present and encoding is not ‘text/plain’, decode the content prior to writing. Default: ‘’
```

but in your local branch we've truncated descriptions leaving us with the less helpful
```
content: (string)
```

we can also quick test these renderd docs locally on the commandline with
```
PYTHONPATH=. python3 -m cloudinit.cmd devel schema --docs all
```

File: doc/rtd/conf.py
Comment (holmanb) [reply doc/rtd/conf.py]: This is fixed in commit  [e1b52e3b6196055](https://github.com/canonical/cloud-init/pull/1101/commits/e1b52e3b619605565311728108334e6da720c42a). I don't think further manipulation of the `__doc__` is required, so these functions can just be removed.

File: doc/rtd/conf.py
Comment (holmanb) [reply doc/rtd/conf.py]: Addressed in [e1b52e3b61960556](https://github.com/canonical/cloud-init/pull/1101/commits/e1b52e3b619605565311728108334e6da720c42a).

File: cloudinit/config/schema.py
Comment (blackboxsw) [inline cloudinit/config/schema.py:536]: So, with this change you have dropped the ability to specify individual schema modules on the commandline via:
```PYTHONPATH=. python3 -m cloudinit.cmd devel schema --docs cc_write_files``` while this may be an under-utilized or under-documented feature of the command line, it does represent a change that we should probably discuss the merit of. 

Minimally, if we are dropping that ability to specify individual modules for documentation on the command line, we need to make sure our [argument parser is updated to avoid prompting for an `--all` value which is now disregarded](https://github.com/canonical/cloud-init/blob/e63069297ba5d75e47c1ed1149dd6cd3f1ac9a92/cloudinit/config/schema.py#L507).

**Review by blackboxsw — CHANGES_REQUESTED** [review]
I think do need to include property descriptions in docs. We might be able to drop the individual module documentation from the command-line, but we need to fixup the argparse setup for the devel schema subcommand.

File: cloudinit/config/schema.py
Comment (holmanb) [reply cloudinit/config/schema.py:124]: Yes, I'll update that to be more obvious. What is getting updated here is actually cloudinitValidator.META_SCHEMA['additionalProperties'].

File: cloudinit/config/schema.py
Comment (holmanb) [reply cloudinit/config/schema.py:413]: And if we ever do get concerned about the performance of python string concatenation, [this might come in handy](https://gist.github.com/holmanb/84be00eab35477565cb95a1d62a741a9).

File: cloudinit/config/schema.py
Comment (holmanb) [reply cloudinit/config/schema.py:145]: @blackboxsw I see what you are getting at. Mutual exclusion is not a requirement. Nothing is stopping us from doing strict metadata validation at runtime. Once this lands and we feel comfortable with how it's working that will be a trivial switch to flip.

For now I've refactored a little for readability and added a test that asserts that a bad schema throws the correct exception. I think the change I just made makes it more obvious _what_ failed when a test fails.

File: cloudinit/config/schema.py
Comment (holmanb) [reply cloudinit/config/schema.py:555]: That command doesn't work for me, but in a container [I see this](https://gist.github.com/holmanb/b69e142a2b6be5729a1f7b7bf4e0434c), which is what I would expect for specifying individual modules to print out. Is there something about running it as a module that causes different behavior perhaps? The logic for selecting which modules to print was moved into `load_doc()` and I think it should still work. The functional change was to concatenate module docstrings rather than call `get_schema_doc()`.

Comment (holmanb) [reply cloudinit/config/schema.py:555]: I just noticed that load_doc() isn't doing the same error checking which causes a behavioral change,. I think that this expected behavior probably warrants a unittest. I'll fix it and add one.

File: tests/unittests/test_cli.py
Comment (holmanb) [reply tests/unittests/test_cli.py:237]: I don't think that issue in the PR is related to this. This test just validates that the supported distros in the `cc_ntp` meta config end up in the final doc. currently that list comes from here:

cloud-init/cloudinit/config/cc_ntp.py
```
distros = ['almalinux', 'alpine', 'centos', 'cloudlinux', 'debian',             
            'eurolinux', 'fedora', 'openEuler', 'opensuse', 'photon',            
            'rhel', 'rocky', 'sles', 'ubuntu', 'virtuozzo']  
```

so `system_info()` is not involved

File: cloudinit/config/schema.py
Comment (holmanb) [reply cloudinit/config/schema.py:560]: @blackboxsw - I just pushed some [tests to verify expected behavior](https://github.com/canonical/cloud-init/pull/1101/commits/662564e19994df3e185d3506a7017bbcd2e30fab).

File: tests/unittests/test_handler/test_schema.py
Comment (holmanb) [inline tests/unittests/test_handler/test_schema.py:96]: While I'm making changes, change this to python package name, rather than distro-specific name.

File: doc/rtd/conf.py
Comment (holmanb) [inline doc/rtd/conf.py:4]: looks like we don't currently lint this file - fix import location

File: cloudinit/config/schema.py
Comment (blackboxsw) [inline cloudinit/config/schema.py:479]: This will also error if additional keys in meta exist beyond what is expected with a message like "Missing module metadata key(s) set()". Is it worth differentiating the errors here a bit?
Note formatting/whitespace might be off in the "suggestion" here.
```suggestion
    error_msg = ""
    if keys.difference(expected):
        error_msg = "Additional unexpected keys found in module meta: {}".format(keys.difference(expected) 
    elif expected.difference(keys):
        error_msg = "Missing expected keys in module meta: {}".format(expected.difference(keys))
    if error_msg:
       raise KeyError(error_msg)
```

**Review by blackboxsw — APPROVED** [review]
Thanks @holmanb any of these remaining comments are nits and I can +1 once you make any changes you think are relevant. Thank you for the restructure, I agree it is easier to read/maintain here and functionally works will with no degradation of existing behavior that I can see.

Approve pending any comments you decide to address and a passing CI run

File: cloudinit/config/schema.py
Comment (holmanb) [reply cloudinit/config/schema.py:463]: This was required on a previous iteration for reasons that are no longer valid. I agree, throwing an exception on empty meta/schema is probably the right approach.

File: cloudinit/config/schema.py
Comment (holmanb) [reply cloudinit/config/schema.py]: factored out to util.py

**Review by blackboxsw — CHANGES_REQUESTED** [review]
Combined diff of my musings for your thoughts. I think we want to minimally make the changes to simplify the get_json_validator return value to a 2-tuple instead of 3-tuple. But other diffs are included in case you find them reasonable.

```diff
diff --git a/cloudinit/config/schema.py b/cloudinit/config/schema.py
index 61f8cd9d..0c46cbdc 100644
--- a/cloudinit/config/schema.py
+++ b/cloudinit/config/schema.py
@@ -100,40 +100,48 @@ def get_jsonschema_validator():
         # assignment-operation pylint warning which appears because this
         # code path isn't written against the latest jsonschema).
         types['string'] = (str, bytes)  # pylint: disable=E1137
+        # disable bottom-level keys in meta-schema
+        strict_metaschema = deepcopymeta_schema=Draft4Validator.META_SCHEMA)
+        strict_metaschema["additionalProperties"] = False
         cloudinitValidator = create(
-            meta_schema=Draft4Validator.META_SCHEMA,
+            meta_schema=strict_metaschema,
             validators=Draft4Validator.VALIDATORS,
             version="draft4",
             default_types=types)
-    return (cloudinitValidator, FormatChecker, None)
+    return (cloudinitValidator, FormatChecker)
 
 
-def validate_cloudconfig_metaschema(schema: dict, throw=True):
+def validate_cloudconfig_metaschema(
+    validator, schema: dict, throw: bool = True
+): 
     """Validate provided schema meets the metaschema definition. Return strict
     Validator and FormatChecker for use in validation
+    @param validator: Draft4Validator instance used to validate the schema
     @param schema: schema to validate
     @param throw: Sometimes the validator and checker are required, even if
-        the schema is invalid. Toggle for whether to raise SchemaErrors
-
-    @returns: Tuple(jsonschema.Validator, jsonschema.FormatChecker, err)
+        the schema is invalid. Toggle for whether to raise
+        SchemaValidationError orlog warnings.
 
     @raises: ImportError when jsonschema is not present
-    @raises: jsonschema.exceptions.SchemaError if the schema is invalid
+    @raises: SchemaValidationError when the schema is invalid
     """
 
     from jsonschema.exceptions import SchemaError
 
-    (cloudinitValidator, FormatChecker, _) = get_jsonschema_validator()
     try:
-
-        # disable bottom-level keys
-        cloudinitValidator.META_SCHEMA["additionalProperties"] = False
-        cloudinitValidator.check_schema(schema)
-    except SchemaError as e:
+        validator.check_schema(schema)
+    except SchemaError as err:
+        # Raise SchemaValidationError so we don't have jsonschema imports
+        # at call sites.
         if throw:
-            raise e
-        return (cloudinitValidator, FormatChecker, e)
-    return (cloudinitValidator, FormatChecker, None)
+            raise SchemaValidationError(
+              errors=(('.'.join([str(p) for p in err.path]), err.message),)
+            )
+        logging.warning(
+            "Meta-schema validation failed, attempting to validate config"
+            " anyway: %s",
+            err
+        )
 
 
 def validate_cloudconfig_schema(
@@ -154,20 +162,14 @@ def validate_cloudconfig_schema(
         against the provided schema.
     """
     try:
-        (cloudinitValidator, FormatChecker, err) = (
-            get_jsonschema_validator()
-            if not strict_metaschema
-            else validate_cloudconfig_metaschema(schema, throw=False)
-        )
+        (cloudinitValidator, FormatChecker) = get_jsonschema_validator()
+        if strict_metaschema:
+            validate_cloudconfig_metaschema(
+                cloudinitValidator, schema, throw=False
+            )
     except ImportError:
         logging.debug("Ignoring schema validation. jsonschema is not present")
         return
-    if err:
-        logging.warning(
-            "Meta-schema validation failed, attempting to validate config"
-            " anyway: %s",
-            err,
-        )
     validator = cloudinitValidator(schema, format_checker=FormatChecker())
     errors = ()
     for error in sorted(validator.iter_errors(config), key=lambda e: e.path):
```

**Comment by holmanb** [comment]
> Combined diff of my musings for your thoughts. I think we want to minimally make the changes to simplify the get_json_validator return value to a 2-tuple instead of 3-tuple. But other diffs are included in case you find them reasonable.

Thanks for the help on this. I wasn't sure about the best approach to the failure paths with this one. Besides one item, I think these are all improvements.

> ```diff
> diff --git a/cloudinit/config/schema.py b/cloudinit/config/schema.py
> index 61f8cd9d..0c46cbdc 100644
> --- a/cloudinit/config/schema.py
> +++ b/cloudinit/config/schema.py
> @@ -100,40 +100,48 @@ def get_jsonschema_validator():
>          # assignment-operation pylint warning which appears because this
>          # code path isn't written against the latest jsonschema).
>          types['string'] = (str, bytes)  # pylint: disable=E1137
> +        # disable bottom-level keys in meta-schema
> +        strict_metaschema = deepcopymeta_schema=Draft4Validator.META_SCHEMA)
> +        strict_metaschema["additionalProperties"] = False
>          cloudinitValidator = create(
> -            meta_schema=Draft4Validator.META_SCHEMA,
> +            meta_schema=strict_metaschema,
>              validators=Draft4Validator.VALIDATORS,
>              version="draft4",
>              default_types=types)
> -    return (cloudinitValidator, FormatChecker, None)
> +    return (cloudinitValidator, FormatChecker)

The problem with this is that get_jsonschema_validator() conditionally produces validator/formatchecker differently based on jsonschema version. The change proposed here only applies to the pre-3.0 jsonschema branch. I think we want both branches to be capable of strict validation. Assuming I don't see anything else in need of change I plan on applying the other proposals you made, in addition to a modified version of this one.

Thanks!

**Review by blackboxsw — APPROVED** [review]
Thank you for all the discussion and work here @holmanb +1. LGTM

**Pull request merged**
